src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.cpp

```diff
@@ -72,5 +72,90 @@ void ThrustCMEstimation::Reset(uint64_t CurrentSimNanos)
  */
 void ThrustCMEstimation::UpdateState(uint64_t CurrentSimNanos)
 {
-    
+    /*! create output message buffers */
+    VehicleConfigMsgPayload vehConfigOutBuffer;
+    CMEstDataMsgPayload cmEstDataBuffer;
+
+    /*! read and allocate the thrustConfigMsg */
+    THRConfigMsgPayload thrConfigBuffer = this->thrusterConfigBInMsg();
+
+    /*! compute thruster information in B-frame coordinates */
+    Eigen::Vector3d r_TB_B = cArray2EigenVector3d(thrConfigBuffer.rThrust_B);
+    Eigen::Vector3d T_B = thrConfigBuffer.maxThrust * cArray2EigenVector3d(thrConfigBuffer.tHatThrust_B);
+
+    /*! compute error w.r.t. target attitude */
+    AttGuidMsgPayload attGuidBuffer = this->attGuidInMsg();
+    Eigen::Vector3d sigma_BR   = cArray2EigenVector3d(attGuidBuffer.sigma_BR);
+    Eigen::Vector3d omega_BR_B = cArray2EigenVector3d(attGuidBuffer.omega_BR_B);
+    double attError = pow(sigma_BR.squaredNorm() + omega_BR_B.squaredNorm(), 0.5);
+
+    /*! read commanded torque msg */
+    CmdTorqueBodyMsgPayload cmdTorqueBuffer = this->intFeedbackTorqueInMsg();
+    Eigen::Vector3d L_B = -cArray2EigenVector3d(cmdTorqueBuffer.torqueRequestBody);
+
+    Eigen::Vector3d y;       // measurement
+    Eigen::Vector3d preFit;  // pre-fit residual
+    Eigen::Vector3d postFit; // post-fit residual
+    Eigen::Matrix3d H;       // linear model
+    Eigen::Matrix3d K;       // kalman gain
+    Eigen::Matrix3d S;
+
+    /*! assign preFit and postFit residuals to NaN, rewrite them in case of measurement update */
+    preFit  = {nan("1"), nan("1"), nan("1")};
+    postFit = {nan("1"), nan("1"), nan("1")};
+
+    if (attError < this->attitudeTol) {
+
+        /*! subtract torque about point B from measurement model */
+        y = L_B - r_TB_B.cross(T_B);
+        /*! H is the skew-symmetric matrix obtained from T_B */
+        H = eigenTilde(T_B);
+        /*! S is defined for convenience */
+        S = H * this->P * H.transpose() + this->R;
+        /*! Kalman gain */
+        K = this->P * H.transpose() * S.inverse();
+
+        /*! pre-fit residuals */
+        preFit = y - H * this->r_CB_est;
+
+        /*! measurement state update */
+        this->r_CB_est = this->r_CB_est + K * preFit;
+        /*! measurement covariance update  */
+        this->P = (this->I - K * H) * this->P;
+
+        /*! post-fit residuals */
+        postFit = y - H * this->r_CB_est;
+    }
+
+    /*! compute state 1-sigma covariance */
+    Eigen::Vector3d sigma;
+    for (int i=0; i<3; i++) {
```

Comment:
:large_blue_circle: A fun thing to note; you could pre increment here, and any other for loop, and save two extra machine instructions (when the loop is compiled). A good explanation of why that works

https://stackoverflow.com/a/4706225

Reply:
So I just change "i++" to "++i"? Done